### ISSUE TITLE
[Performance] Find out why the GPU memory allocated with `CUDAExecutionProvider` is much larger than the ONNX size

### ISSUE THREAD

**Issue by fxmarty** [microsoft/onnxruntime#14526]
### Describe the issue

I have a model that is 4137 MB as a .onnx, exported from a PyTorch's `ScriptModule` through `torch.onnx.export`.

When loading the ONNX model through an InferenceSession using `CUDAExecutionProvider`, **18081 MB** of memory gets allocated on GPU.

When loading the `.pt` weights into a `ScriptModule` on GPU, PyTorch allocates only 4267 MB, which is more reasonable compared to the `.pt` size that is 4267 MB.

Hence, I am wondering how is memory allocation on GPU performed with `CUDAExecutionProvider`, and why it may be more than x4 the ONNX size. It could very well be that my ONNX is ill-formatted, so I'd like to find out where.

### To reproduce

Use this script to find out peak GPU memory allocation (MB):
```sh
a=0
while true; do 
b=$(nvidia-smi --query-gpu=memory.used --format=csv|grep -v memory|awk '{print $1}'| sed -n 2p) # gpu number 1
[ $b -gt $a ] && a=$b && c=$(bc -l <<<"1.048576*$a") && echo $c
sleep .5
done
```

Use this script to load the `ScriptModule`:
```python
import torch
import time

scripted_pipeline = torch.load("scripted_sd_cuda.pt")
time.sleep(3)
```

Use this script to load the ONNX model:
```python
import onnxruntime as ort
import time

providers = ["CUDAExecutionProvider"]

session_options = ort.SessionOptions()
session_options.graph_optimization_level = (
    ort.GraphOptimizationLevel.ORT_ENABLE_BASIC
)

session = ort.InferenceSession("stable_diffusion_pipeline.onnx", providers=providers, sess_options=session_options)
time.sleep(3)
```

### Urgency

normal

### Platform

Linux

### OS Version

Linux 5.15.0-56-generic #62-Ubuntu SMP Tue Nov 22 19:54:14 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.13.1

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

CUDA

### Execution Provider Library Version

CUDA 11.7

### Model File

use:
```
wget https://felix-playground.s3.eu-west-3.amazonaws.com/onnx_model.zip
wget https://felix-playground.s3.eu-west-3.amazonaws.com/stable_diffusion_pipeline.onnx_data
wget https://felix-playground.s3.eu-west-3.amazonaws.com/scripted_sd_cuda.pt
```

The `stable_diffusion_pipeline.onnx_data` is a ~4 GB file with external data.

### Is this a quantized model?

No

**Comment by yuslepukhin** [comment]
CUDA allocations are expensive, so ORT caches them in its own Arena. It attempts to re-use it, but the overall footprint is high.

You can try disabling it with `sessionOptions.DisableCpuMemoryArena()` and see how much your performance depends on it.
CPU is a misnomer as the arena was never meant to be for CPU memory.

To deal with the memory growth, there is a `shrink` feature that may attempt periodically to shrink the arena's memory. It is not perfect.

You can read about it [here](https://onnxruntime.ai/docs/get-started/with-c.html#features).

**Comment by pauldog** [comment]
Same with DirectML. Can use 5-6GB of GPU VRAM plus lots of RAM for about 2GB of onnx files. The DirectML people are saying they are working no reducing the memory size. So cross fingers.

**Comment by fxmarty** [comment]
@yuslepukhin @pauldog Thank you for your advice!

I tried using the CUDA provider options `{"arena_extend_strategy": "kSameAsRequested", "cudnn_conv_algo_search": "HEURISTIC"}`

The

```
run_options = onnxrt.RunOptions()
run_options.add_run_config_entry("memory.enable_memory_arena_shrinkage", "cpu:0;gpu:0")
session.run([], {input_name: x}, run_options)
````

did not help either.

This is a good reference: https://github.com/microsoft/onnxruntime/issues/14038#issuecomment-1368306161

I still don't really understand what's the difference with PyTorch, where the huge memory allocation comes from, and why Pytorch is more memory efficient out-of-the-box. Digging into provider options to have something (maybe) work is not very user friendly. I think I'll give up for now.

**Comment by fxmarty** [comment]
Also noticing a much larger GPU memory usage on the second `.run()` call compared to the first, similar to https://github.com/microsoft/onnxruntime/issues/14477 , I get an out of memory error for the second path but not the first.

1.14 did not help it appears.

**Comment by yuslepukhin** [comment]
Looking at the title of this issue, I am not sure you are looking for a correct thing. Are you comparing ONNX file size with the memory requirements during runtime?

**Comment by fxmarty** [comment]
Yes, roughly. Or better, I am comparing PyTorch's memory usage vs ONNX Runtime on CUDAExecutionProvider.

**Comment by pranavsharma** [comment]
@fxmarty have you tried disabling the arena altogether?

**Comment by tianleiwu** [comment]
@fxmarty, for GPT-j, I saw many Cast nodes from fp16 to fp32:
<img width="253" alt="image" src="https://user-images.githubusercontent.com/30328909/226748984-24dd02c2-e2d8-4efa-a50b-e6eb78f8beaf.png">
That will need more memory to run in fp32.
General guideline is to run graph optimizations first (with CUDA EP), then convert graph to fp16. 

Also, I saw If operator is to used to combine two subgraphs. It is recommended to export two onnx models (one for initial run without past state, another with past state) separately. Those two ONNX models can be used in BeamSearch operator as subgraphs, which will call them in sequential order. In this way, you need not use If operator.

**Comment by fxmarty** [comment]
Thank you for your suggestions @pranavsharma @tianleiwu!

The model was exported with `torch.onnx.export` already with fp16. I did not use ORT fp16 conversion due to the issues I hit with symbolic shape inference we discussed in an other issue.

The thing is that it is not about running the model. It is just about loading it into an InferenceSession:

```python
import onnxruntime as ort

session = ort.InferenceSession("decoder_model_merged.onnx", providers=["CUDAExecutionProvider"])
```

This one allocates 66419 MiB and will OOM if fewer GPU memory than that is available.

Running

```
from transformers import AutoModelForCausalLM
import torch

model = AutoModelForCausalLM.from_pretrained("EleutherAI/gpt-j-6B", torch_dtype=torch.float16).to("cuda")
```

allocates 12157MiB.

I would like to avoid using BeamSearch or GreedySearch operators, as they are I think only supported by CPUExecutionProvider and CUDAExecutionProvider. But you are right that we could support them as an option in the ONNX export - especially if those work with onnxruntime-web.

I will try disabling the arena.

**Comment by lukas-folle-snkeos** [comment]
Is there a way to associate the memory usage per layer and to identify problematic ones?

**Comment by tianleiwu** [comment]
You can build ORT from source and enable node input and outputs dumping, and apply some debug code (or build the branch):
https://github.com/microsoft/onnxruntime/commit/6c0a0aacfdc3b63c99921e7ead69bbeb46dbb3f4
Then you can see the memory allocation.

**Comment by Hung-Ju919** [comment]
@peterukk, @ogencoglu Any finding about your issue? I got a similar issue as you mentioned
